api/src/main/java/org/pmiops/workbench/api/ClusterController.java

```diff
@@ -34,117 +42,109 @@ public Cluster apply(org.pmiops.workbench.notebooks.model.Cluster firecloudClust
-  public ResponseEntity<ClusterListResponse> listClusters(String labels) {
-    List<org.pmiops.workbench.notebooks.model.Cluster> oldClusters;
-    oldClusters = this.notebooksService.listClusters(labels, /* includeDeleted */ false);
-    ClusterListResponse response = new ClusterListResponse();
-    response.setItems(oldClusters.stream().map(TO_ALL_OF_US_CLUSTER).collect(Collectors.toList()));
-    return ResponseEntity.ok(response);
+  private org.pmiops.workbench.notebooks.model.ClusterRequest createFirecloudClusterRequest() {
+    org.pmiops.workbench.notebooks.model.ClusterRequest firecloudClusterRequest = new org.pmiops.workbench.notebooks.model.ClusterRequest();
+    // TODO(calbach): Configure Jupyter server extensions here.
+    Map<String, String> labels = new HashMap<String, String>();
+    labels.put("all-of-us", "true");
+    labels.put("created-by", userProvider.get().getEmail());
+    firecloudClusterRequest.setLabels(labels);
+    return firecloudClusterRequest;
   }
 
   @Override
-  public ResponseEntity<Void> localizeNotebook(String workspaceNamespace, String workspaceId,
-      List<FileDetail> fileList) {
-    String clusterName = convertClusterName(workspaceId);
-    this.notebooksService.localize(workspaceNamespace, clusterName,
-        convertfileDetailsToMap(fileList));
+  public ResponseEntity<Void> localize(
+      String projectName, String clusterName, ClusterLocalizeRequest body) {
+    String workspaceBucket;
+    try {
+      workspaceBucket = "gs://" + fireCloudService.getWorkspace(body.getWorkspaceNamespace(), body.getWorkspaceId())
+          .getWorkspace()
+          .getBucketName();
+    } catch (ApiException e) {
+      if (e.getCode() == 404) {
+        log.log(Level.INFO, "Firecloud workspace not found", e);
+        throw new NotFoundException(String.format(
+            "workspace %s/%s not found or not accessible",
+            body.getWorkspaceNamespace(), body.getWorkspaceId()));
+      }
+      throw ExceptionUtils.convertFirecloudException(e);
+    }
+
+    // For the common case where the notebook cluster matches the workspace
+    // namespace, simply name the directory as the workspace ID; else we
+    // include the namespace in the directory name to avoid possible conflicts
+    // in workspace IDs.
+    String workspacePath = body.getWorkspaceId();
+    if (!projectName.equals(body.getWorkspaceNamespace())) {
+      workspacePath = body.getWorkspaceNamespace() + ":" + body.getWorkspaceId();
+    }
+    String localDir = String.join("/", "~", "workspaces", workspacePath);
+    Map<String, String> toLocalize = body.getNotebookNames()
+        .stream()
+        .collect(Collectors.<String, String, String>toMap(
+            name -> localDir + "/" + name,
+            name -> String.join("/", workspaceBucket, "notebooks", name)));
+    // TODO(calbach): Localize a delocalize config JSON file as well, once Leo supports this.
+    toLocalize.put(
+        localDir + "/.all_of_us_config.json",
```

Comment:
Is the leading dot supposed to be here before "all"?

FWIW, the latest code in pyclient no longer looks for this config file in ~/config; it looks for it in config/, relative to the notebook. It seems like we'll either need to update this code or pyclient to match.

Reply:
ALL_OF_US_CONFIG_PATH normally won't be set in notebook environments; it's there to allow an override for running locally / during integration tests, here:

https://github.com/all-of-us/pyclient/blob/d41e3f2bcbeebf36daeba825a93dea3474e82daa/libproject/devstart.rb#L52

When this environment variable is not set, the path defaults to:

config/all_of_us_config.json

as shown here:

https://github.com/all-of-us/pyclient/blob/master/py/aou_workbench_client/config.py#L10

You'll need to change that in a separate PR to be .all_of_us_config.json, I suppose, and make a new tag for a pyclient, then modifying the startup script to use it here:

https://github.com/all-of-us/workbench/blob/master/api/scripts/setup_notebook_cluster.sh

Having multiple config files under multiple subdirs should be fine, provided that each config file is in the same path relative to the notebook file referencing it.

Reply:
Ah I see now, it's CWD-relative. OK, I think this should work, thanks.

Will prep a PR and send before I merge this.